eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/DualFundingHandlers.scala

```diff
@@ -60,62 +61,20 @@ trait DualFundingHandlers extends CommonFundingHandlers {
-        // ERR_INFORMATION_LEAK, because here we receive a response from the watcher from a WatchConfirmed that we put
-        // ourselves and somehow forgot.
-        log.error(s"internal error: the funding tx that confirmed doesn't match any of our funding txs: ${fundingTx.bin}")
-        None
-    }
-  }
-
-  def acceptDualFundingTx(d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED, fundingTx: Transaction, realScidStatus: RealScidStatus): Option[(DATA_WAIT_FOR_DUAL_FUNDING_READY, ChannelReady)] = {
-    pruneCommitments(d.metaCommitments, fundingTx).map {
-      metaCommitments =>
-        // after pruning, there is only one commitments
-        watchFundingTx(metaCommitments.commitments.head)
-        realScidStatus match {
-          case _: RealScidStatus.Temporary => context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, fundingTx))
-          case _ => () // zero-conf channel
-        }
-        val shortIds = createShortIds(d.channelId, realScidStatus)
-        val channelReady = createChannelReady(shortIds, metaCommitments.params)
-        d.deferred.foreach(self ! _)
-        (DATA_WAIT_FOR_DUAL_FUNDING_READY(metaCommitments, shortIds), channelReady)
-    }
-  }
-
-  /**
-   * If one of the funding txs confirms while we are disconnected, we need to put a watch on its output, because our
-   * counterparty could force-close using a previous commitment. We can also rollback and forget deprecated funding
-   * transactions immediately.
-   */
-  def handleDualFundingConfirmedOffline(w: WatchFundingConfirmedTriggered, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED) = {
-    pruneCommitments(d.metaCommitments, w.tx) match {
-      case Some(metaCommitments) =>
-        // after pruning, there is only one commitments
-        watchFundingTx(metaCommitments.commitments.head)
-        context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
-        if (d.previousFundingTxs.nonEmpty) {
-          log.info(s"funding txid={} was confirmed in state disconnected, cleaning up {} alternative txs", w.tx.txid, d.metaCommitments.commitments.size - 1)
-        }
-        stay() using d.copy(metaCommitments = metaCommitments) storing()
-      case None =>
-        stay()
+  def acceptFundingTxConfirmed(w: WatchFundingConfirmedTriggered, d: PersistentChannelData): Option[MetaCommitments] = {
+    log.info("funding txid={} was confirmed at blockHeight={} txIndex={}", w.tx.txid, w.blockHeight, w.txIndex)
+    val fundingStatus = ConfirmedFundingTx(w.tx)
+    context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
+    d.metaCommitments.updateLocalFundingStatus(w.tx.txid, fundingStatus).map { metaCommitments1 =>
+      // first of all, we watch the funding tx that is now confirmed
+      val commitments = metaCommitments1.commitments.find(_.fundingTxId == w.tx.txid).get // TODO: rework this
```

Comment:
I like the fact that we return an explicit result, because the way it is used it should always succeed. Maybe we should make it return an `Either`, which would be consistent with other `MetaCommitments` methods.

Reply:
Done with [8c9c0ab](https://github.com/ACINQ/eclair/pull/2598/commits/8c9c0abaa819f01883cc3d96f4c96109a84114e6).